sdks/python/apache_beam/coders/coder_impl.py

```diff
@@ -578,17 +588,44 @@ def __init__(self, value_coder, timestamp_coder, window_coder):
 
   def encode_to_stream(self, value, out, nested):
     wv = value  # type cast
-    self._value_coder.encode_to_stream(wv.value, out, True)
     # Avoid creation of Timestamp object.
-    out.write_bigendian_int64(wv.timestamp_micros)
+    restore_sign = -1 if wv.timestamp_micros < 0 else 1
+    out.write_bigendian_uint64(
+        # Convert to postive number and divide, since python rounds off to the
+        # lower negative number. For ex: -3 / 2 = -2, but we expect it to be -1,
+        # to be consistent across SDKs.
+        self._from_normal_time(
+            restore_sign * (abs(wv.timestamp_micros) / 1000)))
     self._windows_coder.encode_to_stream(wv.windows, out, True)
+    # Default PaneInfo encoded byte representing NO_FIRING.
+    # TODO(vikasrk): Remove the hard coding here once PaneInfo is supported.
+    out.write_byte(0xF)
+    self._value_coder.encode_to_stream(wv.value, out, nested)
 
   def decode_from_stream(self, in_stream, nested):
+    timestamp = self._to_normal_time(in_stream.read_bigendian_uint64())
+    # Restore MIN/MAX timestamps to their actual values as encoding incurs loss
+    # of precision while converting to millis.
+    # Note: This is only a best effort here as there is no way to know if these
+    # were indeed MIN/MAX timestamps.
+    # TODO(vikasrk): Clean this up once we have a BEAM wide consensus on
+    # precision of timestamps.
+    if timestamp == -(abs(MIN_TIMESTAMP.micros) / 1000):
```

Comment:
Do we want to do <= and >= here and below?  This is somewhat fragile.  Can you also document the Java min and max values?